// common/Page/PageEntryManagement/Comps/Checkbox/index.js
/**
 * @file index.js
 * @author gyl
 */
import React from 'react';
import {extendObservable, observable, runInAction} from 'mobx';
import {observer} from 'mobx-react';

@observer
export default class Checkbox extends React.Component {
    handleChange(e) {
        const comp = this.props.comp;

        comp.toggleChecked(e.target.checked);
        comp.clearError();
    }

    validate(e) {
        const comp = this.props.comp;

        if (comp.validate) {
            comp.validate();
        }
    }

    _setIndeterminate() {
        const comp = this.props.comp;

        if (comp.state.partial) {
            this.refs.checkbox.indeterminate = true;
        }
    }

    componentDidMount() {
        this._setIndeterminate();
    }

    componentDidUpdate() {
        this._setIndeterminate();
    }

    render() {
        const comp = this.props.comp;

        // console.log('@debug, ', this.props)

        const checkProps = {
            checked: comp.state.checked,
            disabled: comp.state.disabled
        };

        return <div>
            <input value={comp.state.id}
                   type="checkbox"
                   ref="checkbox"
                {...checkProps}
                   onChange={e => this.handleChange(e)}
                   onBlur={e => this.validate(e)}/>
            <label>{comp.state.dispLabel}</label>
            {comp.state.hasError ? comp.state.errorMessage : null}
        </div>;
    }
}
