// common/Page/PageEntryManagement/Comps/Checkbox/model.js
/**
 * @file modal.js
 * @author gyl
 */
import {observable, action} from 'mobx';
import _ from 'lodash';

export default class CheckboxModel {
    @observable state = {
        type: 'checkbox',

        // true | false
        checked: false,
        partial: true,
        disabled: false,

        extra: null,

        id: '',
        dispLabel: '',

        // @todo 用于校验
        hasError: false,
        errorMessage: ''
    }

    constructor(checked, id, dispLabel, opts = {}) {
        const {
            partial,
            extra,
            disabled
        } = opts;

        this.initValue(checked, id, dispLabel, partial, extra, disabled);
    }

    @action initValue(checked, id, dispLabel, partial, extra, disabled) {
        _.extend(this.state, {
            checked, id, dispLabel, partial, extra, disabled
        });
    }

    @action toggleChecked(checked) {
        if (arguments.length === 0) {
            this.state.checked = !this.state.checked;
        }
        else {
            this.state.checked = checked;
        }
    }

    @action clearError() {
        this.state.hasError = false;
    }

    wrapData() {
        return {
            id: this.state.id,
            checked: this.state.checked,
            disabled: this.state.disabled,
            extra: this.state.extra
        };
    }
}
