// common/components/ErrorWall/index.js

import React from 'react'
import PropTypes from 'prop-types'
import {h, c} from 'utils/base'

import compStyle from './style.use.less'

export default class ErrorWall extends React.Component {
    static propTypes = {
        classPrefix: PropTypes.string,
        /*
         * 错误类型，不同类型使用不同背景图
         * default: 'default'
         * */
        type: PropTypes.oneOf(['default', '404', '403', '500', 'plain']),
        /*
         * ** deprecate **
         * 改用`props.children`作为的内容
         * 图片下方的文案
         * default: ''
         * */
        content: PropTypes.node
    }
    static defaultProps = {
        classPrefix: 'erp',
        type: 'default',
        content: ''
    }

    componentWillMount() {
        compStyle.use()
    }

    componentWillUnmount() {
        compStyle.unuse()
    }

    render() {
        const {type, content, children, className, ...props} = this.props
        const compClass = this.compClass
        const classString = c(
            compClass,
            className,
        )

        return h.div(classString, {},
            h.div(`${compClass}-type-${type}`),
            h.div(`${compClass}-content`, {},
                children || content
            )
        )
    }

    get compClass() {
        return `${this.props.classPrefix}-error-wall`
    }
}
