// common/Page/PageEntryManagement/Page/state.js
import React from 'react';
import {extendObservable, computed, observable, action} from 'mobx';
import _ from 'lodash';
import agent from 'utils/ajax-agent';
import urlUtils from 'utils/lib/url-utils';

import CheckboxModel from '../Comps/Checkbox/model';

class EntryPage {
  @observable pages = []
  @observable checkboxes = []

  @action initPages(entryPages) {
    _.each(entryPages, (config, key) => {
      this.pages.push({
        key,
        ...config
      });
    });

    this.checkboxes = this.pages.map(pageConfig =>
      new CheckboxModel(pageConfig.turnedOn, pageConfig.key,
        <span>{pageConfig.key}
          [<a href={urlUtils.link('', '', pageConfig.key)} target='_blank'>跳转至该模块</a>]
                </span>,
        {disabled: pageConfig.openOnDefault}
      )
    );
  }

  submit() {
    let entries = this.checkboxes.map(checkbox => checkbox.wrapData());
    // console.log('@debug, ', entries);

    agent.devPost('/toggle-entries', {data: {entries}});
  }


  fetchEntries() {
    agent.devPost('/all-entries')
      .then(response => {
        this.initPages(response.data);
      });
  }
}

export function mount(page) {
  page.entryPage = new EntryPage();
}

export function unmount(page) {
  delete page.entryPage;
}
